src/ts/Connection.ts

```diff
@@ -0,0 +1,481 @@
+	DbActionSubroutineInputTypes,
+	DbFeatureResponseTypes,
+	DbSubroutineInputOptionsMap,
+	DbSubroutineResponseTypes,
+	DbSubroutineResponseTypesMap,
+	GenericObject,
+	Logger,
+} from '#shared/types';
+import { dummyLogger } from '#shared/utils';
+import { dependencies as serverDependencies } from '../.mvomrc.json';
+import compileModel, { type ModelConstructor } from './compileModel';
+import type Schema from './Schema';
+
+// #region Types
+export interface CreateConnectionOptions {
+	/** Optional logger instance */
+	logger?: Logger;
+	/**
+	 * Lifetime of cache of db server data (s)
+	 * @defaultValue 3600
+	 */
+	cacheMaxAge?: number;
+	/**
+	 * Request timeout (ms)
+	 * 0 implies no timeout
+	 * @defaultValue 0
+	 */
+	timeout?: number;
+}
+
+export enum ConnectionStatus {
+	// convert to enum when transitioning class to TS
+	disconnected = 'disconnected',
+	connected = 'connected',
+	connecting = 'connecting',
+}
+export interface DeployFeaturesOptions {
+	/**
+	 * Create directory when deploying features
+	 * @defaultValue false
+	 */
+	createDir?: boolean;
+}
+
+type ServerDependency = keyof typeof serverDependencies;
+
+interface ServerFeatureSet {
+	validFeatures: Map<ServerDependency, string>;
+	invalidFeatures: Set<ServerDependency>;
+}
+// #endregion
+
+/** A connection object */
+class Connection {
+	/** File system path of the UniBasic source code */
+	private static unibasicPath = path.resolve(path.join(__dirname, '../', 'unibasic'));
+
+	/** Connection status */
+	public status: ConnectionStatus = ConnectionStatus.disconnected;
+
+	/** Logger instance used for diagnostic logging */
+	public logger: Logger;
+
+	/** Object providing the current state of db server features and availability */
+	private serverFeatureSet: ServerFeatureSet = {
+		validFeatures: new Map(),
+		invalidFeatures: new Set(),
+	};
+
+	/** Time that the connection information cache will expire */
+	private cacheExpiry = 0;
+
+	/** Maximum age of the cache before it must be refreshed */
+	private cacheMaxAge: number;
+
+	/** URI of the full endpoint for communicating with the database */
+	private endpoint: string;
+
+	/** +/- in milliseconds between database server time and local server time */
+	private timeDrift?: number;
+
+	/** Request timeout, in milliseconds */
+	private timeout: number;
+
+	private constructor(
+		/** URI of the MVIS which facilitates access to the mv database */
+		mvisUri: string,
+		/** Database account that connection will be used against */
+		account: string,
+		/** Logger instance */
+		logger: Logger,
+		/** Lifetime of cache of db server data (s) */
+		cacheMaxAge: number,
+		/** Request timeout (ms) */
+		timeout: number,
+	) {
+		this.logger = logger;
+		this.cacheMaxAge = cacheMaxAge;
+		this.endpoint = `${mvisUri}/${account}/subroutine/${Connection.getServerProgramName('entry')}`;
+		this.timeout = timeout;
+
+		logger.debug(`creating new connection instance`);
+	}
+
+	public static createConnection(
```

Comment:
This is now a static factory method rather than a standalone method.  The constructor is private so the only way to instantiate this should be via the factory.